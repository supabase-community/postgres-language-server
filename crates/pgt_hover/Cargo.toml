[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_hover"
repository.workspace = true
version              = "0.0.0"


[dependencies]
humansize                  = { version = "2.1.3" }
pgt_query.workspace        = true
pgt_schema_cache.workspace = true
pgt_text_size.workspace    = true
pgt_treesitter.workspace   = true
schemars                   = { workspace = true, optional = true }
serde                      = { workspace = true, features = ["derive"] }
serde_json                 = { workspace = true }
sqlx.workspace             = true
tokio                      = { version = "1.41.1", features = ["full"] }
tracing                    = { workspace = true }
tree-sitter.workspace      = true
tree_sitter_sql.workspace  = true

[dev-dependencies]
insta                    = { version = "1.42.1" }
pgt_test_utils.workspace = true

[lib]
doctest = false

[features]
schema = ["dep:schemars"]
