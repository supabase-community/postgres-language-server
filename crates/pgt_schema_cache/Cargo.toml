[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_schema_cache"
repository.workspace = true
version              = "0.0.0"


[dependencies]
anyhow.workspace          = true
async-std                 = { version = "1.12.0" }
futures-util              = "0.3.31"
pgt_console.workspace     = true
pgt_diagnostics.workspace = true
serde.workspace           = true
serde_json.workspace      = true
sqlx.workspace            = true
strum                     = { workspace = true }
tokio.workspace           = true

[dev-dependencies]
pgt_test_utils.workspace = true

[lib]
doctest = false
