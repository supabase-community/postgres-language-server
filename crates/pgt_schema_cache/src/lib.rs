//! The schema cache

#![allow(dead_code)]

mod columns;
mod functions;
mod policies;
mod roles;
mod schema_cache;
mod schemas;
mod tables;
mod triggers;
mod types;
mod versions;

pub use columns::*;
pub use functions::{Behavior, Function, FunctionArg, FunctionArgs, ProcKind};
pub use policies::{Policy, PolicyCommand};
pub use roles::*;
pub use schema_cache::SchemaCache;
pub use schemas::Schema;
pub use tables::{ReplicaIdentity, Table, TableKind};
pub use triggers::{Trigger, TriggerAffected, TriggerEvent};
pub use types::{PostgresType, PostgresTypeAttribute};
