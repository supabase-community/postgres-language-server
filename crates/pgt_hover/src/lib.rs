use pgt_schema_cache::SchemaCache;
use pgt_text_size::TextSize;
use pgt_treesitter::TreeSitterContextParams;

use crate::{
    contextual_priority::prioritize_by_context, hoverables::Hoverable, hovered_node::HoveredNode,
    to_markdown::format_hover_markdown,
};

mod contextual_priority;
mod hoverables;
mod hovered_node;
mod to_markdown;

pub struct OnHoverParams<'a> {
    pub position: TextSize,
    pub schema_cache: &'a SchemaCache,
    pub stmt_sql: &'a str,
    pub ast: Option<&'a pgt_query::NodeEnum>,
    pub ts_tree: &'a tree_sitter::Tree,
}

#[tracing::instrument(level = "debug", skip_all, fields(
    text = params.stmt_sql,
    position = params.position.to_string()
))]
pub fn on_hover(params: OnHoverParams) -> Vec<String> {
    let ctx = pgt_treesitter::context::TreesitterContext::new(TreeSitterContextParams {
        position: params.position,
        text: params.stmt_sql,
        tree: params.ts_tree,
    });

    if let Some(hovered_node) = HoveredNode::get(&ctx) {
        let items: Vec<Hoverable> = match hovered_node {
            HoveredNode::Table(node_identification) => match node_identification {
                hovered_node::NodeIdentification::Name(n) => params
                    .schema_cache
                    .find_tables(n.as_str(), None)
                    .into_iter()
                    .map(Hoverable::from)
                    .collect(),

                hovered_node::NodeIdentification::SchemaAndName((s, n)) => params
                    .schema_cache
                    .find_tables(n.as_str(), Some(&s))
                    .into_iter()
                    .map(Hoverable::from)
                    .collect(),

                _ => vec![],
            },

            HoveredNode::Column(node_identification) => match node_identification {
                hovered_node::NodeIdentification::Name(column_name) => params
                    .schema_cache
                    .find_cols(&column_name, None, None)
                    .into_iter()
                    .map(Hoverable::from)
                    .collect(),

                hovered_node::NodeIdentification::SchemaAndName((table_or_alias, column_name)) => {
                    // resolve alias to actual table name if needed
                    let actual_table = ctx
                        .get_mentioned_table_for_alias(table_or_alias.as_str())
                        .map(|s| s.as_str())
                        .unwrap_or(table_or_alias.as_str());

                    params
                        .schema_cache
                        .find_cols(&column_name, Some(actual_table), None)
                        .into_iter()
                        .map(Hoverable::from)
                        .collect()
                }

                _ => vec![],
            },

            HoveredNode::Function(node_identification) => match node_identification {
                hovered_node::NodeIdentification::Name(function_name) => params
                    .schema_cache
                    .find_functions(&function_name, None)
                    .into_iter()
                    .map(Hoverable::from)
                    .collect(),

                hovered_node::NodeIdentification::SchemaAndName((schema, function_name)) => params
                    .schema_cache
                    .find_functions(&function_name, Some(&schema))
                    .into_iter()
                    .map(Hoverable::from)
                    .collect(),

                _ => vec![],
            },

            HoveredNode::Role(node_identification) => match node_identification {
                hovered_node::NodeIdentification::Name(role_name) => params
                    .schema_cache
                    .find_roles(&role_name)
                    .into_iter()
                    .map(Hoverable::from)
                    .collect(),

                _ => vec![],
            },

            HoveredNode::Schema(node_identification) => match node_identification {
                hovered_node::NodeIdentification::Name(schema_name) => params
                    .schema_cache
                    .find_schema(&schema_name)
                    .map(Hoverable::from)
                    .map(|s| vec![s])
                    .unwrap_or_default(),

                _ => vec![],
            },

            HoveredNode::PostgresType(node_identification) => match node_identification {
                hovered_node::NodeIdentification::Name(type_name) => params
                    .schema_cache
                    .find_type(&type_name, None)
                    .map(Hoverable::from)
                    .map(|s| vec![s])
                    .unwrap_or_default(),

                hovered_node::NodeIdentification::SchemaAndName((schema, type_name)) => params
                    .schema_cache
                    .find_type(&type_name, Some(schema.as_str()))
                    .map(Hoverable::from)
                    .map(|s| vec![s])
                    .unwrap_or_default(),

                _ => vec![],
            },

            _ => todo!(),
        };

        prioritize_by_context(items, &ctx)
            .into_iter()
            .map(|item| format_hover_markdown(&item, params.schema_cache))
            .filter_map(Result::ok)
            .collect()
    } else {
        Default::default()
    }
}
