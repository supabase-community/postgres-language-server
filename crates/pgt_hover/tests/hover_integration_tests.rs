use pgt_hover::{OnHoverParams, on_hover};
use pgt_schema_cache::SchemaCache;
use pgt_test_utils::QueryWithCursorPosition;
use pgt_text_size::TextSize;
use sqlx::{Executor, PgPool};

async fn test_hover_at_cursor(name: &str, query: String, setup: Option<&str>, test_db: &PgPool) {
    if let Some(setup) = setup {
        test_db
            .execute(setup)
            .await
            .expect("Failed to setup test database");
    }

    let schema_cache = SchemaCache::load(test_db)
        .await
        .expect("Failed to load Schema Cache");

    let (position, sql) = QueryWithCursorPosition::from(query).get_text_and_position();

    let mut parser = tree_sitter::Parser::new();
    parser
        .set_language(tree_sitter_sql::language())
        .expect("Error loading sql language");

    let tree = parser.parse(&sql, None).unwrap();
    let ast = pgt_query::parse(&sql)
        .ok()
        .map(|parsed| parsed.into_root().unwrap());

    let hover_results = on_hover(OnHoverParams {
        position: TextSize::new(position as u32),
        schema_cache: &schema_cache,
        stmt_sql: &sql,
        ast: ast.as_ref(),
        ts_tree: &tree,
    });

    let mut snapshot = String::new();
    snapshot.push_str("# Input\n");
    snapshot.push_str("```sql\n");
    snapshot.push_str(&sql);
    snapshot.push('\n');

    for _ in 0..position {
        snapshot.push(' ');
    }
    snapshot.push_str("â†‘ hovered here\n");

    snapshot.push_str("```\n\n");

    if hover_results.is_empty() {
        snapshot.push_str("# Hover Results\n");
        snapshot.push_str("No hover information found.\n");
    } else {
        snapshot.push_str("# Hover Results\n");
        for (i, result) in hover_results.iter().enumerate() {
            if i > 0 {
                snapshot.push_str("\n---\n\n");
            }
            snapshot.push_str(result);
            snapshot.push('\n');
        }
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(name, snapshot);
    });
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_function_hover_builtin_count(test_db: PgPool) {
    let query = format!(
        "select cou{}nt(*) from users",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("builtin_count", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_function_hover_builtin_now(test_db: PgPool) {
    let query = format!(
        "select n{}ow() from users",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("builtin_now", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_function_hover_builtin_max(test_db: PgPool) {
    let query = format!(
        "select m{}ax(id) from users",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("builtin_max", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_function_hover_custom_function(test_db: PgPool) {
    let setup = r#"
        create or replace function custom_add(a integer, b integer)
        returns integer
        language plpgsql
        immutable
        as $$
        begin
            return a + b;
        end;
        $$;
    "#;

    let query = format!(
        "select custom_a{}dd(1, 2)",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("custom_function", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_function_hover_with_schema(test_db: PgPool) {
    let setup = r#"
        create schema test_schema;
        create or replace function test_schema.schema_func(text)
        returns text
        language sql
        stable
        as $$
            select $1 || ' processed';
        $$;
    "#;

    let query = format!(
        "select test_schema.schema_f{}unc('test')",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("function_with_schema", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_column_hover_with_table_ref(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            email varchar(255) not null
        );
    "#;

    let query = format!(
        "select users.i{}d from users",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("column_hover", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_column_hover_in_join(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            email varchar(255) not null
        );

        create table posts (
            id serial primary key,
            user_id serial references users(id),
            content text
        );
    "#;

    let query = format!(
        "select * from users u join posts p on u.id = p.use{}r_id",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("column_hover_join", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_table_hover_works(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            email varchar(255) not null
        );
    "#;

    let query = format!(
        "select id from use{}rs",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("table_hover", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_no_hover_on_keyword(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            email varchar(255) not null
        );
    "#;

    let query = format!(
        "sel{}ect id from users",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("no_hover_keyword", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn shortens_lengthy_functions(test_db: PgPool) {
    let setup = r#"
        create or replace function public.func(cool_stuff text,       something_else int,a_third_thing text)
        returns void
        language sql
        stable
        as $$
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
            select 1;
        $$;
    "#;

    let query = format!(
        "select public.fu{}nc()",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("lenghty_function", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_role_hover_create_role(test_db: PgPool) {
    let query = format!(
        "create role alternate_ow{}ner with superuser createdb login",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("role_create", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_role_hover_alter_role(test_db: PgPool) {
    let query = format!(
        "alter role test_log{}in set work_mem = '256MB'",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("role_alter", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_table_hover_with_quoted_schema(test_db: PgPool) {
    let setup = r#"
        create schema auth;
        create table auth.users (
            id serial primary key,
            email varchar(255) not null
        );
    "#;

    let query = format!(
        r#"select * from "auth".use{}rs"#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("table_hover_quoted_schema", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_function_hover_with_quoted_schema(test_db: PgPool) {
    let setup = r#"
        create schema auth;
        create or replace function auth.authenticate_user(user_email text)
        returns boolean
        language plpgsql
        security definer
        as $$
        begin
            return exists(select 1 from auth.users where email = user_email);
        end;
        $$;
    "#;

    let query = format!(
        r#"select "auth".authenticate_u{}ser('test@example.com')"#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("function_hover_quoted_schema", query, Some(setup), &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_column_hover_with_quoted_schema_table(test_db: PgPool) {
    let setup = r#"
        create schema auth;
        create table auth.user_profiles (
            id serial primary key,
            user_id int not null,
            first_name varchar(100),
            last_name varchar(100)
        );
    "#;

    let query = format!(
        r#"select "auth"."user_profiles".first_n{}ame from "auth"."user_profiles""#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor(
        "column_hover_quoted_schema_table",
        query,
        Some(setup),
        &test_db,
    )
    .await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_table_hover_with_quoted_table_name(test_db: PgPool) {
    let setup = r#"
        create schema auth;
        create table auth.users (
            id serial primary key,
            email varchar(255) not null
        );
    "#;

    let query = format!(
        r#"select * from "auth"."use{}rs""#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor(
        "table_hover_quoted_table_name",
        query,
        Some(setup),
        &test_db,
    )
    .await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_column_hover_with_quoted_column_name(test_db: PgPool) {
    let setup = r#"
        create schema auth;
        create table auth.users (
            id serial primary key,
            email varchar(255) not null
        );
    "#;

    let query = format!(
        r#"select "ema{}il" from auth.users"#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor(
        "column_hover_quoted_column_name",
        query,
        Some(setup),
        &test_db,
    )
    .await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_column_hover_with_quoted_column_name_with_table(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            email varchar(255) not null
        );

        create table phone_nums (
            phone_id serial primary key,
            email varchar(255) not null,
            phone int
        );
    "#;

    let query = format!(
        r#"select phone, id from users join phone_nums on "users"."em{}ail" = phone_nums.email;"#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor(
        "column_hover_quoted_column_name_with_table",
        query,
        Some(setup),
        &test_db,
    )
    .await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_policy_table_hover(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            name text
        );
    "#;

    test_db.execute(setup).await.unwrap();

    let query = format!(
        r#"create policy "my cool pol" on us{}ers for all to public with check (true);"#,
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("create_policy", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_revoke_table_hover(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            name text
        );
    "#;

    test_db.execute(setup).await.unwrap();

    let query = format!(
        "revoke select on us{}ers from public;",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("revoke_select", query, None, &test_db).await;
}

#[sqlx::test(migrator = "pgt_test_utils::MIGRATIONS")]
async fn test_grant_table_hover(test_db: PgPool) {
    let setup = r#"
        create table users (
            id serial primary key,
            name text
        );
    "#;

    test_db.execute(setup).await.unwrap();

    let query = format!(
        "grant select on us{}ers to public;",
        QueryWithCursorPosition::cursor_marker()
    );

    test_hover_at_cursor("grant_select", query, None, &test_db).await;
}
