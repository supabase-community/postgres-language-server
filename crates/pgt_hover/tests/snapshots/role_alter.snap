---
source: crates/pgt_hover/tests/hover_integration_tests.rs
expression: snapshot
---
# Input
```sql
alter role test_login set work_mem = '256MB'
                   â†‘ hovered here
```

# Hover Results
### `test_login` - ðŸ‘¤ login  
#### Login role  
---
